grpc/controlproxy/controlproxy.go

```diff
@@ -0,0 +1,245 @@
+	req := &rpc.ValidateDepsRequest{
+		Metrics: common.NewMetrics(mts),
+		Plugins: common.SubPluginsToMsg(plugins),
+	}
+	reply, err := c.Client.ValidateDeps(context.Background(), req)
+	serrs := groupErrors(err, reply.Errors)
+	return serrs
+}
+
+func (c ControlProxy) SubscribeDeps(taskID string, mts []core.Metric, plugins []core.Plugin) []serror.SnapError {
+	req := depsRequest(taskID, mts, plugins)
+	reply, err := c.Client.SubscribeDeps(context.Background(), req)
+	serrs := groupErrors(err, reply.Errors)
+	return serrs
+}
+
+func (c ControlProxy) UnsubscribeDeps(taskID string, mts []core.Metric, plugins []core.Plugin) []serror.SnapError {
+	req := depsRequest(taskID, mts, plugins)
+	reply, err := c.Client.UnsubscribeDeps(context.Background(), req)
+	serrs := groupErrors(err, reply.Errors)
+	return serrs
+}
+
+func (c ControlProxy) MatchQueryToNamespaces(namespace []string) ([][]string, serror.SnapError) {
+	req := &rpc.ExpandWildcardsRequest{
+		Namespace: namespace,
+	}
+	reply, err := c.Client.MatchQueryToNamespaces(context.Background(), req)
+	if err != nil {
+		return nil, serror.New(err)
+	}
+	if reply.Error != nil {
+		return nil, common.ToSnapError(reply.Error)
+	}
+	nss := ToNSS(reply.NSS)
+	return nss, nil
+}
+
+///---------Util-------------------------------------------------------------------------
+// TODO(CDR): move the functions used by outside callers only to a util package and make functions used only
+// in this file non-exported.
+func getPluginType(t core.PluginType) int32 {
+	val := int32(-1)
+	switch t {
+	case core.CollectorPluginType:
+		val = 0
+	case core.ProcessorPluginType:
+		val = 1
+	case core.PublisherPluginType:
+		val = 2
+	}
+	return val
+}
+func groupErrors(err error, errs []*common.SnapError) []serror.SnapError {
+	serrs := make([]serror.SnapError, 0)
+	if err != nil {
+		serrs = append(serrs, serror.New(err))
+	}
+	serrs = append(serrs, common.ConvertSnapErrors(errs)...)
+	return serrs
+}
+
+func depsRequest(taskID string, mts []core.Metric, plugins []core.Plugin) *rpc.SubscribeDepsRequest {
+	req := &rpc.SubscribeDepsRequest{
+		Metrics: common.NewMetrics(mts),
+		Plugins: common.CorePluginsToMsg(plugins),
+		TaskId:  taskID,
+	}
+	return req
+}
+
+func ConvertNSS(nss [][]string) []*rpc.ArrString {
+	res := make([]*rpc.ArrString, len(nss))
+	for i := range nss {
+		var tmp rpc.ArrString
+		tmp.S = nss[i]
+		res[i] = &tmp
+	}
+	return res
+}
+
+func ToNSS(arr []*rpc.ArrString) [][]string {
+	nss := make([][]string, len(arr))
+	for i, v := range arr {
+		nss[i] = v.S
+	}
+	return nss
+}
+
+func ErrorsToStrings(in []error) []string {
+	if len(in) == 0 {
+		return []string{}
+	}
+	erro := make([]string, len(in))
+	for i, e := range in {
+		erro[i] = e.Error()
+	}
+	return erro
+}
+
+func ReplyErrorsToErrors(errs []string) []error {
+	if len(errs) == 0 {
+		return []error{}
+	}
+	erro := make([]error, len(errs))
```

Comment:
This is a bug. It needs to be either `var erro []error` since it's using append, or change it to assign to the index directly.